sql/binlog.cc

```diff
@@ -5577,13 +5577,19 @@ void MYSQL_BIN_LOG::report_cache_write_error(THD *thd, bool is_transactional) {
   }
 
 #ifdef WITH_WSREP
-  /* If wsrep transaction is active and binlog emulation is on,
-     binlog write error may leave transaction without any registered
-     htons. This makes wsrep rollback hooks to be skipped and the
-     transaction will remain alive in wsrep world after rollback.
-     Register binlog hton here to ensure that rollback happens in full. */
+  /*
+     If wsrep transaction is active and binlog emulation is on, binlog
+     write error may leave transaction without any registered htons.
+     This makes wsrep rollback hooks to be skipped and the transaction
+     will remain alive in wsrep world after rollback.  Register binlog
+     hton here to ensure that rollback happens in full only when the
+     session is in multi stmt transaction mode (i.e, either
+     autocommit=OFF or when the transaction is started with explicit
+     BEGIN).
+  */
   if (WSREP_EMULATE_BINLOG(thd)) {
-    if (is_transactional) trans_register_ha(thd, true, binlog_hton, NULL);
+    if (is_transactional && thd->in_multi_stmt_transaction_mode())
+      trans_register_ha(thd, true, binlog_hton, NULL);
     trans_register_ha(thd, false, binlog_hton, NULL);
```

Comment:
should there be else here
if (is_transactional && multi) {
   trans_register(.. true ..) 
} else {
  trans_register_ha(..flase);
}

Reply:
In case of a muli-stmt transaction, both statement and transaction scoped binlog handlers should be registered.
Having (if..else) only registers only one scope and may cause undefined behavior. 

https://github.com/percona/percona-xtradb-cluster/blob/8.0/sql/binlog.cc#L10136-L10137 also does the same.

Reply:
Right venky, I remember now